Kubelet get pod from HTTP source

chentao1596 opened the pull request:
why we need it:
Only support one header parameter when getting pod from HTTP source, it's not work very well when server need multi header parameters.

What this PR does:
1) Add one run parameter for kubelet: "manifest-url-headers" - to be used to supporting more than one HTTP header
2) Change some auto-generated codes

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37539)
<!-- Reviewable:end -->

Release note：
[release-note.txt](https://github.com/kubernetes/kubernetes/files/624092/release-note.txt)

chentao11596 committed:
add new run parameter manifest-url-headers

chentao11596 committed:
regenerated codes

dims commented:
@k8s-bot ok to test

chentao11596 committed:
add flags for verify

mtaufen commented on cmd/kubelet/app/options/options.go:
```diff
@@ -98,6 +98,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&s.HTTPCheckFrequency.Duration, "http-check-frequency", s.HTTPCheckFrequency.Duration, "Duration between checking http for new data")
 	fs.StringVar(&s.ManifestURL, "manifest-url", s.ManifestURL, "URL for accessing the container manifest")
 	fs.StringVar(&s.ManifestURLHeader, "manifest-url-header", s.ManifestURLHeader, "HTTP header to use when accessing the manifest URL, with the key separated from the value with a ':', as in 'key:value'")
+	fs.StringSliceVar(&s.ManifestURLHeaders, "manifest-url-headers", s.ManifestURLHeaders, "Comma-separated list of HTTP headers to use when accessing the manifest URL, with the key separated from the value with a ':', as in 'key:value', will be combined with ManifestURLHeader")
```
I think we should deprecate `ManifestURLHeader` in favor of `ManifestURLHeaders` and mark the `manifest-url-header` flag as deprecated with `fs.MarkDeprecated(...)`. It seems kind of redundant to keep both of these around for an extended period of time, since the new flag can accomplish the job of the old flag (ofc the old flag still needs to work for a few releases for backwards compatibility reasons).

@mikedanese what do you think?

mtaufen commented on pkg/apis/componentconfig/types.go:
```diff
@@ -137,6 +137,10 @@ type KubeletConfiguration struct {
+	// manifestURLHeaders is a comma-separated list of HTTP headers to use
+	// when accessing the manifest URL, with the key separated from the value
+	// with a ':', as in 'key:value', will be combined with ManifestURLHeader
+	ManifestURLHeaders []string `json:"manifestURLHeaders"`
```
nit: /s/`, will be combined with ManifestURLHeader`/`. These will be applied in addition to the header provided via ManifestURLHeader.`

mtaufen commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -191,6 +191,10 @@ type KubeletConfiguration struct {
+	// manifestURLHeaders is a comma-separated list of HTTP headers to use
+	// when accessing the manifest URL, with the key separated from the value
+	// with a ':', as in 'key:value', will be combined with ManifestURLHeader
+	ManifestURLHeaders []string `json:"manifestURLHeaders"`
```
nit: /s/`, will be combined with ManifestURLHeader`/`. These will be applied in addition to the header provided via ManifestURLHeader.`

mtaufen commented:
Thanks for this! Mostly looks good other than a few nits and thoughts on deprecating the old flag. This will also require a release note since it makes changes to Kubelet flags.

chentao1596 commented:
@k8s-ci-robot I signed it!

chentao1596 commented:
@mtaufen hi, I've changed it, PTAL, thx for your work!

chentao11596 committed:
re-autogenerate codes

regenerated codes

add flags for verify

resolve conflicts

resolve conflict

resolve conflicts

chentao11596 committed:
recover before resolve conflicts

chentao1596 commented:
@mtaufen 
hi, i am very sorry that i prepare to close this PR because of i can't resolve conflicts under this branch, i have create the other PR, can you help me look it ? thank you!
https://github.com/kubernetes/kubernetes/pull/38089
